### ISSUE TITLE
Allow public API definition via annotations

### ISSUE THREAD

**Issue by busbey** [revelc/apilyzer-maven-plugin#20]
I'd like to start using this on HBase (and maybe Hadoop eventually). Both of those projects define their public API via annotations.

Coming up with regexes that are equivalent would be labor intensive and brittle. I believe the annotations are scoped RUNTIME, so finding them should be doable.

**Comment by ctubbsii** [comment]
This is a good idea... @keith-turner and I had discussed this, and I'm surprised we hadn't already opened an issue for it.

**Comment by keith-turner** [comment]
Anyway this would be a really cool feature.  I also think it would be fairly simple to implement.

I had looked into doing this for hadoop and had come to the conclusion that the annotations were not carried forward to the class.   However I am not sure what I was looking at.   Looking at [InterfaceAudience.java](https://github.com/apache/hadoop/blob/release-2.6.0/hadoop-common-project/hadoop-annotations/src/main/java/org/apache/hadoop/classification/InterfaceAudience.java), the retention is Runtime.

**Comment by keith-turner** [comment]
Ok I figured out where I went wrong.   I was looking at hadoop 2.2.0 source because that what Accumulo depends on and it was esy to look at that in Eclipse.  Hadoop 2.2.0 [InterfaceAudience.java](https://github.com/apache/hadoop/blob/release-2.2.0/hadoop-common-project/hadoop-annotations/src/main/java/org/apache/hadoop/classification/InterfaceAudience.java) does not have runtime scope.

**Comment by keith-turner** [comment]
I poked at this a little bit and AFAICT the hadoop 2.6.0 jars in maven central do not have the annotations.   Not sure if something is stripping them during the hadoop build or if I was doing something wrong in my code.  I will push the branch with my experimental changes later today.

**Comment by busbey** [comment]
Ping when the experimental changes are up; I'd like to test on HBase.

**Comment by keith-turner** [comment]
@busbey I pushed the branch to https://github.com/keith-turner/accumulo-api-analyzer/tree/apilyzer-20

Could add a test pom for hbase like the one i added for hadoop in that branch.

**Comment by busbey** [comment]
okay I added one. the test output looks like it's not picking up any instanced of the annotation there as well. :disappointed:

**Comment by keith-turner** [comment]
> okay I added one. the test output looks like it's not picking up any instanced of the annotation there as well.

I am still not 100% there is not some bug in my code.   Was thinking of producing my own little test jar with an annotated class.  

If annotations are being stripped in hadoop and hbase build, its possible that the plugin may work as part of the hadoop and hbase build if it executes before whatever strips.

**Comment by busbey** [comment]
AFAIK HBase at least has nothing purposefully stripping. We define the annotations themselves as a part of our API, so I would expect them to be present.

I'll take a look at the class files.

**Comment by busbey** [comment]
Well, it's present in the class file. Take a look at [this javap for HBase's Table class](https://gist.github.com/busbey/368c48256fc62a4d2ca1)

     RuntimeVisibleAnnotations:
        0: #8()
        1: #9()

and then in the constant table:

     #8 = Utf8               Lorg/apache/hadoop/hbase/classification/InterfaceAudience$Public;
     #9 = Utf8               Lorg/apache/hadoop/hbase/classification/InterfaceStability$Evolving;

**Comment by keith-turner** [comment]
I just took another look at this.   I used javap to look at hadoop's Configuration class downloaded via maven.  I am also seeing the RuntimeVisibleAnnotations stuff there.  This indicates that hadoop is not removing anything via post processing.

```bash
MAVEN_REPO_LOC=<your maven repo location>
javap -v -c -cp $MAVEN_REPO_LOC/org/apache/hadoop/hadoop-common/2.7.0/hadoop-common-2.7.0.jar org.apache.hadoop.conf.Configuration
```
Below are similar snippets to what was in the hbase class.    `RuntimeVisibleAnnotations` seems to be a top level element in the javap output.   

```
   #992 = Class              #1457        // org/apache/hadoop/classification/InterfaceAudience$Public
   #994 = Utf8               Lorg/apache/hadoop/classification/InterfaceAudience$Public;
   #995 = Class              #1458        // org/apache/hadoop/classification/InterfaceStability$Stable
   #997 = Utf8               Lorg/apache/hadoop/classification/InterfaceStability$Stable;
```

```
RuntimeVisibleAnnotations:
  0: #994()
  1: #997()
```

**Comment by keith-turner** [comment]
I figured out the problem with my branch.  Stupid mistake on my part.  I was calling `ClassInfo.getClass()` instead of `ClassInfo.load()`.  Now I am running into some other problem.

**Comment by keith-turner** [comment]
I got something working and pushed a commit to my branch.  Here is a [Gist](https://gist.github.com/keith-turner/ae8d447438aee65e4c1d) with some output for Hadoop and HBase.  I updated the HBase pom, not entirely sure its correct.  Also, I think the Hadoop test pom needs more deps to analyze the entire HAdoop API.

There is something goofy going on with `org.apache.hadoop.hbase.HConstants$Modify` in the hbase output.  I will look into that further.  Its neat to see it working.  Found things like `FileSystem listStatusIterator()` returning `RemoteIterator`.  `FileSystem` is Public+Stable where `RemoteIterator` is Public+Evolving.  Ideally Public+Stable types would only reference other Public+Stable types.  Found similar things for HBase.
